File: utils/downloadModels.js
#!/usr/bin/env node

/**
 * Download Models from HuggingFace Hub on Startup
 * Ensures full functionality even when deployed
 */

const https = require('https');
const fs = require('fs');
const path = require('path');

const MODELS = [
    {
        name: 'efficientnetb0_notop.h5',
        size: '16.7 MB',
        // Use public model hub URL - will be set after upload
        url: 'https://huggingface.co/huyhuy07/leaf-disease-detector-1/resolve/main/efficientnetb0_notop.h5'
    },
    {
        name: 'plant_disease_model.h5',
        size: '39.7 MB',
        url: 'https://huggingface.co/huyhuy07/leaf-disease-detector-1/resolve/main/plant_disease_model.h5'
    },
    {
        name: 'mango_model.h5',
        size: '49.1 MB',
        url: 'https://huggingface.co/huyhuy07/leaf-disease-detector-1/resolve/main/mango_model.h5'
    }
];

const MODEL_DIR = path.join(__dirname, '../model');

// Ensure model directory exists
if (!fs.existsSync(MODEL_DIR)) {
    fs.mkdirSync(MODEL_DIR, { recursive: true });
}

async function downloadFile(url, filepath, filename) {
    return new Promise((resolve, reject) => {
        console.log(`📥 Downloading ${filename}...`);

        const file = fs.createWriteStream(filepath);
        const request = https.get(url, (response) => {
            if (response.statusCode !== 200) {
                reject(new Error(`Download failed: ${response.statusCode}`));
                return;
            }

            const len = parseInt(response.headers['content-length'], 10);
            let downloaded = 0;

            response.on('data', (chunk) => {
                downloaded += chunk.length;
                const percent = Math.round((downloaded / len) * 100);
                process.stdout.write(`\r  Progress: ${percent}%`);
            });

            response.pipe(file);

            file.on('finish', () => {
                file.close();
                console.log(`\n✅ ${filename} downloaded!`);
                resolve();
            });

            file.on('error', (err) => {
                fs.unlink(filepath, () => { });
                reject(err);
            });
        });

        request.on('error', reject);
    });
}

async function ensureModelsExist() {
    console.log(`
╔════════════════════════════════════════╗
║   🤖 LOADING AI MODELS                 ║
╚════════════════════════════════════════╝
  `);

    for (const model of MODELS) {
        const modelPath = path.join(MODEL_DIR, model.name);

        if (fs.existsSync(modelPath)) {
            const stats = fs.statSync(modelPath);
            const size = (stats.size / (1024 * 1024)).toFixed(1);
            console.log(`✅ ${model.name} (${size} MB) - Already loaded`);
        } else {
            try {
                await downloadFile(model.url, modelPath, model.name);
            } catch (error) {
                console.error(`\n⚠️  Warning: Could not download ${model.name}`);
                console.error(`   Error: ${error.message}`);
                console.log(`   Make sure to upload models manually or configure URLs\n`);
            }
        }
    }

    console.log(`
╔════════════════════════════════════════╗
║   ✅ MODELS READY!                    ║
╚════════════════════════════════════════╝
  `);
}

// Export for use in server
module.exports = { ensureModelsExist };

// Run if called directly
if (require.main === module) {
    ensureModelsExist().catch((err) => {
        console.error('Failed to ensure models:', err);
        process.exit(1);
    });
}

File: model/download_models.py
#!/usr/bin/env python3
"""
Download all 6 H5 models from HuggingFace Hub on startup
"""

import os
import sys
from huggingface_hub import hf_hub_download

HF_USERNAME = "huyhuy07"
REPO_NAME = "leaf-disease-detector-models"
MODEL_DIR = os.path.join(os.path.dirname(__file__), '..', 'model')

# Create model directory if not exists
os.makedirs(MODEL_DIR, exist_ok=True)

MODELS = [
    'efficientnet_merged.h5',
    'efficientnetb0_notop.h5',
    'leaf_disease_model.h5',
    'leaf_disease_modhel.h5',
    'mango_model.h5',
    'plant_disease_model.h5'
]

def download_models():
    """Download all models from HuggingFace"""
    
    print("=" * 60)
    print("📥 DOWNLOADING 6 H5 MODELS FROM HUGGINGFACE")
    print("=" * 60)
    print()
    
    downloaded = 0
    skipped = 0
    
    for model_file in MODELS:
        model_path = os.path.join(MODEL_DIR, model_file)
        
        # Skip if already exists
        if os.path.exists(model_path):
            size_mb = os.path.getsize(model_path) / (1024 * 1024)
            print(f"✓ {model_file} ({size_mb:.2f} MB) - already cached")
            skipped += 1
            continue
        
        try:
            print(f"📥 Downloading {model_file}...", end=" ", flush=True)
            
            hf_hub_download(
                repo_id=f"{HF_USERNAME}/{REPO_NAME}",
                filename=model_file,
                local_dir=MODEL_DIR,
                repo_type="model"
            )
            
            size_mb = os.path.getsize(model_path) / (1024 * 1024)
            print(f"✓ ({size_mb:.2f} MB)")
            downloaded += 1
            
        except Exception as e:
            print(f"❌ Error: {e}")
            return False
    
    print()
    print("=" * 60)
    print(f"✅ Download Complete:")
    print(f"   Downloaded: {downloaded}")
    print(f"   Cached: {skipped}")
    print(f"   Total: {len(MODELS)}")
    print("=" * 60)
    
    return True

if __name__ == "__main__":
    success = download_models()
    sys.exit(0 if success else 1)
